src/ethereum_test_specs/helpers.py

```diff
@@ -91,63 +122,112 @@ def __init__(
+    def __init__(
+        self,
+        *,
+        execution_context: ExecutionContext,
+        want_exception: List[ExceptionBase] | ExceptionBase | None,
+        got_exception: ExceptionWithMessage | UndefinedException | None,
+        context: Dict[str, Any],
+    ):
+        """Initialize the exception."""
+        self.execution_context = execution_context
+        self.want_exception = want_exception
+        self.got_exception = (
+            got_exception.exception
+            if isinstance(got_exception, ExceptionWithMessage)
+            else got_exception
         )
-    elif expected_error and info.t8n_error_message:
-        if isinstance(info.tx.error, List):
-            for expected_exception in info.tx.error:
-                expected_error_msg = exception_mapper.exception_to_message(expected_exception)
-                if expected_error_msg is None:
-                    continue
-                if expected_error_msg in info.t8n_error_message:
-                    # One of expected exceptions is found in tx error string, no error
-                    return
-
-        if isinstance(info.tx.error, List):
-            expected_exception = info.tx.error[0]
-        elif info.tx.error is None:
-            return  # will never happen but removes python logic check
+        if self.got_exception is None:
+            self.got_message = None
         else:
-            expected_exception = info.tx.error
-
-        expected_error_msg = exception_mapper.exception_to_message(expected_exception)
-        t8n_error_exception = exception_mapper.message_to_exception(info.t8n_error_message)
-        exception_mapper_name = exception_mapper.__class__.__name__
-
-        if expected_error_msg is None or expected_error_msg not in info.t8n_error_message:
-            raise TransactionExceptionMismatchError(
-                index=info.transaction_index,
-                nonce=info.tx.nonce,
-                expected_exception=expected_exception,
-                expected_message=expected_error_msg,
-                got_exception=t8n_error_exception,
-                got_message=info.t8n_error_message,
-                mapper_name=exception_mapper_name,
+            self.got_message = (
+                got_exception.message
+                if isinstance(got_exception, ExceptionWithMessage)
+                else str(got_exception)
+            )
+        self.context = context
+
+    def verify(self: "ExceptionInfo", *, strict_match: bool) -> None:
+        """Verify the exception."""
+        want_exception, got_exception = (
+            self.want_exception,
+            self.got_exception,
+        )
+        if want_exception and not got_exception:
+            raise UnexpectedExecutionSuccessError(
+                execution_context=self.execution_context, **self.context
+            )
+        elif not want_exception and got_exception:
+            assert self.got_message is not None
```

Comment:
I think this will be the case if `isinstance(got_exception, UndefinedException)`? 

If that's the case, as `UnexpectedExecutionFailError` is probably the most interesting aspect of the fail and we should avoid raising an (unhelpful) assertion error before reporting it?

Reply:
This is covered by:
```python
        if self.got_exception is None:
            ...
        else:
            self.got_message = (
                got_exception.message
                if isinstance(got_exception, ExceptionWithMessage)
                else str(got_exception)
            )
```
`self.got_exception` is not `None`, and also it's not `ExceptionWithMessage`, so it will be a `UndefinedException` converted to `str`.